### PR TITLE
Improve exported router types

### LINKED ISSUE #7764: [typings] Router push and replace methods supporting UrlLike objects
# Bug report

## Describe the bug

After upgrading to next canary (future v9) from 8.1.1, I got a few TypeScript errors when calling `Router.push(url)` and `Router.replace(url)` with object-like URLs.

This seems like a regression in built-in typings compared to the third-party ones (`@types/*`). Related to https://github.com/zeit/next.js/pull/6644.

## To Reproduce

```tsx
import { Router, useRouter } from "next/router";

const router: Router = useRouter();

// 👍
router.push("/about");
router.replace("/about");

// ❌ Argument of type '{ pathname: string; }' is not assignable to parameter of type 'string'.ts(2345)
router.push({ pathname: "/about" });
router.replace({ pathname: "/about" });
```

## Expected behavior

I expected url-like objects to be still acceptable. Not sure if new typings are simply incomplete or objects are arguments are being deprecated.

https://github.com/zeit/next.js/blob/cc497568b4b226498069c9a7adbdc03e689b8526/packages/next-server/lib/router/router.ts#L208-L226

Typings in `@types/next-server`:

* [Definition of `UrlLike` type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/132720a17e15cdfcffade54dd4a23a21c1e16831/types/next-server/router.d.ts#L19
)
* [Usage in `push() and` `replace()`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/132720a17e15cdfcffade54dd4a23a21c1e16831/types/next-server/router.d.ts#L48-L58)

There exists a local `Url` type in `next/client`, which seems like what's needed. Should it be moved to `next-server/router`?

https://github.com/zeit/next.js/blob/b60985b2be964dcac436d025a1943ded6454deaf/packages/next/client/link.tsx#L24

It's already [used in `<Link />` prop types](https://github.com/zeit/next.js/blob/b60985b2be964dcac436d025a1943ded6454deaf/packages/next/client/link.tsx#L49-L50).

---

One more thing: `useRouter()` returns `any` because of how  `RouterContext` is typed 🤔
https://github.com/zeit/next.js/blob/cc497568b4b226498069c9a7adbdc03e689b8526/packages/next/client/router.ts#L125-L127
https://github.com/zeit/next.js/blob/b1fdffec75a7a7ce24693e4e1974714ac7733551/packages/next-server/lib/router-context.ts#L3

## System information

- Version of Next.js: `8.1.1-canary.69`

### PR DESCRIPTION
**Pull request by lfades**
Using the changes from #7768

- the `PublicRouterInstance` was moved to `next-server` and added to multiple places to properly match the actual methods that can be used, instead of the `Router` type which is part of the internal router that can't be used outside our code base.

Fixes #7764

### CODE REVIEW

File: packages/next/lib/data.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export function createHook(
     throw new Error('key not provided to createHook options.')
   }
   return function useData(...args: Array<string | number>) {
-    const router: import('next-server/lib/router/router').default = useContext(
+    const router: import('next-server/lib/router/router').PublicRouterInstance = useContext(
```
Comment (tills13) [inline packages/next/lib/data.ts]: Just curious why this is written this way?

Comment (lfades) [reply packages/next/lib/data.ts]: To avoid importing the file at all, in this case we only need the type.

Comment (tills13) [reply packages/next/lib/data.ts]: iirc, TypeScript strips out empty import lines (e.g. those whose imports are all types)

Comment (lfades) [reply packages/next/lib/data.ts]: That's all I can tell you haha, I can't really appreciate another difference here

Comment (tills13) [reply packages/next/lib/data.ts]: :+1: just looked out of place considering all the other imports are at the top. Doesn't matter as far as I know.

**Review by huv1k — APPROVED** [review]

**Review by Timer — CHANGES_REQUESTED** [review]
We can't let this increase bundle sizes. Should we just rename `PublicRouterInstance` to `NextRouter` and call it a day?

**Review by Timer — CHANGES_REQUESTED** [review]
See above

**Review by Timer — APPROVED** [review]
LGTM if tests pass. We've concluded the size stats must be incorrect because these changes are 100% TypeScript types only.

**Pull request merged**